### ISSUE TITLE
Nushell is printing (or asking to print) escape sequences

### ISSUE THREAD

**Issue by mrkkrp** [nushell/nushell#11950]
### Describe the bug

Please see https://github.com/alacritty/alacritty/issues/7762 for context (at first I thought it was a problem with alacritty), in particular this comment https://github.com/alacritty/alacritty/issues/7762#issuecomment-1961654503. After a recent upgrade I sometimes see weird escape sequences and the output of commands is truncated, e.g.:

```
mark@pad ~> cal
^[[28;16R    February 2024
Mo Tu We Th Fr Sa Su
          1  2  3  4
 5  6  7  8  9 10 11
12 13 14 15 16 17 18
mark@pad ~> cal
^[[29;16R    February 2024
Mo Tu We Th Fr Sa Su
          1  2  3  4
 5  6  7  8  9 10 11
12 13 14 15 16 17 18
19 20 21 22 23 24 25
mark@pad ~> cal
^[[30;16R    February 2024
Mo Tu We Th Fr Sa Su
          1  2  3  4
 5  6  7  8  9 10 11
12 13 14 15 16 17 18
19 20 21 22 23 24 25
26 27 28 29
mark@pad ~> cal
^[[31;16R    February 2024
Mo Tu We Th Fr Sa Su
          1  2  3  4
 5  6  7  8  9 10 11
12 13 14 15 16 17 18
19 20 21 22 23 24 25
26 27 28 29

mark@pad ~> cal
^[[32;16R    February 2024
Mo Tu We Th Fr Sa Su
          1  2  3  4
 5  6  7  8  9 10 11
12 13 14 15 16 17 18
19 20 21 22 23 24 25
26 27 28 29
```

This seems to begin undeterministically, I do not know what exactly provokes it.

### How to reproduce

I do not know.

### Expected behavior

No weird escape sequences should be printed and the output of commands should not be truncated.

### Screenshots

![nushell](https://github.com/nushell/nushell/assets/8165792/3b7ce9be-6226-46d4-9564-043cb8bd06c9)

### Configuration

```
^[[32;60R| key                | value                                                             |
| ------------------ | ----------------------------------------------------------------- |
| version            | 0.89.0                                                            |
| branch             |                                                                   |
| commit_hash        |                                                                   |
| build_os           | linux-x86_64                                                      |
| build_target       | x86_64-unknown-linux-gnu                                          |
| rust_version       | rustc 1.75.0 (82e1608df 2023-12-21) (built from a source tarball) |
| cargo_version      | cargo 1.75.0                                                      |
| build_time         | 1980-01-01 00:00:00 +00:00                                        |
| build_rust_channel | release                                                           |
| allocator          | mimalloc                                                          |
| features           | default, sqlite, trash, which, zip                                |
| installed_plugins  |                                                                   |
```

**Comment by fdncred** [comment]
I'm trying to unravel what's going on here. This is what I get with the nushell `cal` command.
![image](https://github.com/nushell/nushell/assets/343840/eaa0427a-72b5-4218-91e9-faa0793a4be8)

**Comment by sophiajt** [comment]
@fdncred you might need to use `^cal` to make sure you're calling the external `cal`

**Comment by sophiajt** [comment]
I tried to repro this on macOS using the macOS external `cal` and a recent alacritty and wasn't able to. It's possible the issue is limited to Linux?

**Comment by mrkkrp** [comment]
Indeed I have this in my `config.nu`:

```nushell
alias cal = ^cal -m
```

**Comment by NotTheDr01ds** [comment]
@fdncred  It's in the linux-utils package, I believe.

**Comment by NotTheDr01ds** [comment]
@mrkkrp Anything special in your prompt?

**Comment by NotTheDr01ds** [comment]
Ah yeah, this brings back memories of trying to get Alacritty running in WSL with Nix ... I never was successful.  It's not Alacritty's fault - Nix + OpenGL doesn't play well, IIRC.  Or maybe it was Nix + OpenGL + WSL.

**Comment by fdncred** [comment]
found one with `ncal` that works.

**Comment by fdncred** [comment]
@mrkkrp I'm struggling to reproduce this problem. Any tips?

**Comment by mrkkrp** [comment]
I only know as much as this comment says: https://github.com/alacritty/alacritty/issues/7762#issuecomment-1961654503.
It is not about `cal` and I don't know the exact conditions that cause this, but it happens *eventually* and pretty often.

**Comment by mrkkrp** [comment]
Looks like some kind of race condition where either cleaning up happens before the escape sequence is printed (too early) or it happens after everything is printed (too late).

**Comment by ZhichaoHong** [comment]
This happens to me frequently. It get worse if I used oh-my-posh prompt.

**Comment by fdncred** [comment]
I think this is typically a terminal or 3rd party app responsibility. What's happening here is a program is asking for the cursor position with ansi escape `6n` and the terminal is responding with the location like `^[[line:columnR`. Something has to read stdin in order to understand the response. So, if `cal` is asking for the position `cal` should be reading the output from the terminal.

This could also be a bug in crossterm or how we've implemented it in nushell since crossterm "speaks" ansi escape sequences too.

**Comment by ZhichaoHong** [comment]
The problem happens when I set nushell as the default (not login) shell for kitty. If I started nushell from another shell (e.g. zsh), the problem does not happen.

**Comment by mrkkrp** [comment]
FYI, the problem is present regardless of the terminal emulator used. Originally I observed it with Alacritty, now I sometimes get it with wezterm (more rarely though). This seems to indicate that the problem is in Nushell, somehow. It does not happen with other shells.

**Comment by noverby** [comment]
@mrkkrp did you find a solution or workaround?

**Comment by mrkkrp** [comment]
@noverby I'm not seeing it all that often anymore (switching to wezterm seems to have helped a lot).

**Comment by MMukundi** [comment]
Hi all! Is there anyone working on this? The issue still seems somewhat ambiguous, but I'm still running into this pretty often, so I'd be happy to help work on this however I can

**Comment by noverby** [comment]
I have had this issue both in Wezterm and VSCode xterm.js

**Comment by hustcer** [comment]
> I think this is a terminal problem. I haven't been able to reproduce it at all on Windows with WT or on MacOS with Ghostty or WezTerm.

I had this issue on macOS with wezterm, it's not quite often and hard to reproduce

**Comment by qfel** [comment]
+1 for Guake; not sure how to reproduce, but once it happens it keeps happening till I restart nu - and it usually starts happening fairly quickly. Seems to have nothing to do with the actual command though - even pressing enter will cause something like `^[[7;20R` to be displayed under prompt.

**Comment by salo-dea** [comment]
I have a similar issue that only appeared in 0.98.0 on Windows (WindowsTerminal)
For me it seems to happen only when using some old cygwin GNU coreutils (6.10), and even when ignoring their output.
E.g. `cat --version` or  `cat --version | ignore` will break the prompt.
Same commands in powershell or nushell 0.97.1 are fine

This ends up looking especially bad with the default config, I guess due to the right prompt

default config (stays broken):
![image](https://github.com/user-attachments/assets/1c2ccffd-860e-4f1c-a82f-6b3b41d52c21)

my usual config that doesn't have a right prompt (fine again for the next command):
![image](https://github.com/user-attachments/assets/769b9349-48dd-4a63-8c3b-a3423aada3ef)

**Comment by fdncred** [comment]
it looks like something is breaking crossterm and leaving it in raw mode. i'm not sure if that's a reedline thing or nushell nor how to fix it. it would be good if someone could dig in and figure out what's going on. i know this is a showstopper for some.

**Comment by salo-dea** [comment]
I did a git bisect between 0.97.1 and 0.98.0 -> my issue seems to appear at a88f46c6c9ad19867304fc812f81b99ee8e0094f

Disabling the right prompt doesn't help. I suppose there are some other escapes being printed...
![image](https://github.com/user-attachments/assets/2daa5c9e-cc6e-43c5-a203-21249f643b9e)

**Comment by fdncred** [comment]
That's interesting. Is the only way to reproduce it with a cygwin cat command?

**Comment by fdncred** [comment]
I wonder if this version is old enough where escape sequences weren't even supported in windows so cygwin was doing something to make up for that?

**Comment by fdncred** [comment]
I added a revert PR for anyone who wants to try it to see if it fixes their issue here https://github.com/nushell/nushell/pull/13959

**Comment by fdncred** [comment]
This is an alternate PR that would be good for y'all to test https://github.com/nushell/nushell/pull/13961

**Comment by salo-dea** [comment]
@fdncred 
I tried both PR's and they both seem to fix my issue.

In case you need to do some additional test, I zipped my local cat.exe + needed dll's here:
[old_cat.zip](https://github.com/user-attachments/files/17192135/old_cat.zip)
https://www.virustotal.com/gui/file/d2288469162875ca0bbe0f6b704768a9e40e6bc305e9ee6aa0dcabe9167ba6d1/relations

I guess in general my problem is probably unrelated to the initial problem in this thread.. sorry for hijacking it

**Comment by MatrixManAtYrService** [comment]
I've switched my daily driver to 166d5fa4ffa3c5514b2d7597f3238e6828c986f9 (latest main at time of this writing, includes https://github.com/nushell/nushell/pull/13961).  🤞 it doesn't come back up... (I'll report back either way)

**Comment by MatrixManAtYrService** [comment]
I don't think #13961 has fixed it, note the `^[[3;25R` below:

```nushell
❯  echo 'foo' | grep bar
^[[3;25RError: nu::shell::non_zero_exit_code

  × External command had a non-zero exit code
   ╭─[entry #104:1:14]
 1 │ echo 'foo' | grep bar
   ·              ──┬─
   ·                ╰── exited with code 1
   ╰────
```

**Comment by noverby** [comment]
@MatrixManAtYrService do you ase Windows?

**Comment by MatrixManAtYrService** [comment]
No, this is on MacOs (installed via home-manager, running inside of zellij)

**Comment by fdncred** [comment]
@MatrixManAtYrService https://github.com/nushell/nushell/pull/13961 is a windows change, so if you're not on windows it shouldn't do anything.

**Comment by MatrixManAtYrService** [comment]
Oh I see, re: it being a windows change.  Oops 😳 

I have tried to replicate it by just resizing in various ways and running simple commands like `ls`, bit I haven't had luck.  It would seem that there's something in my typical workflow (or maybe it just takes time) which triggers it in context with resizes, but resizes alone don't do it.

**Comment by justinlovinger** [comment]
I also have this issue, running Zellij on Foot. The problem tends to get worse over time, and Nushell eventually crashes.

I also have the issue with Zellij running over Mosh.

Frustratingly, whenever I _try_ to reproduce the issue, it eludes me. That is also why I have not been able to confirm whether the issue occurs on Foot alone, without Zellij. As far as I can tell, the issue seems to occur more often when running commands with large outputs and when often resizing the terminal.

**Comment by MatrixManAtYrService** [comment]
I found a way to reliably (:crossed_fingers:) replicate this.  Or at least something quite similar to this. I've done it on two machines:

- NixOS w/ nushell `0.101.0`
- MacOS w/ nushwll `0.99.1`

I made a gist for it: https://gist.github.com/MatrixManAtYrService/636b9b964314122c2b7b4fe76defebc0

I'm not sure what is special about the way that `pyright` is handling its output, but as long as it finds enough errors, this happens when I pipe the output to `head`.

**Comment by fdncred** [comment]
Are you saying that when you're in other shells like bash/fish/zsh and you do these same commands with pyright, you don't get the same results as you do in nushell (visible ansi escape sequences)?

It looks to me like pyright is printing with ansi escape sequences for color and head doesn't know what to do with it and pyright leaves the terminal in raw mode. Just a guess.

**Comment by MatrixManAtYrService** [comment]
Other shells have shown me ansi escape sequences before in certain situations (though I haven't known pyright to be a particular offender). For example they'll get stuck in a certain color because I've managed to truncate something that restores sanity. But they don't continue to inject new ones with every subsequent command. After this happens once, all subsequent commands begin misbehaving until I close the shell.

Also, I don't think that `^[[57;10R` is an escape sequence but rather the response to a cursor position request (row 57 column 10). Under normal operation, command output can be as long as it wants to, and after the command finishes, nushell (or I guess reedline) asks the terminal where the cursor is so it can redraw the prompt in the right place. It listens for the terminal's response and and draws it such that the previous command output is still visible.  This pyright trick appears to break that and it stays broken even when I go run `echo` or other well-behaved commands.

pyright detects that stdout is not a terminal and disables coloring, so I don't think there are escape sequences in here. My suspicion is that it's more to do with the amount of output and the timing of things.  I suspect this because it doesn't happen with just one or two errors (pyright finds and indicates errors), there needs to be at least four (though it's more reliable with 100), and it also doesn't happen with cat, just head. If I save the output to a file and then redirect to head, no trouble.

**Comment by justinlovinger** [comment]
> I found a way to reliably ( 🤞 ) replicate this. Or at least something quite similar to this. I've done it on two machines:
> 
> * NixOS w/ nushell `0.101.0`
> * MacOS w/ nushwll `0.99.1`
> 
> I made a gist for it: https://gist.github.com/MatrixManAtYrService/636b9b964314122c2b7b4fe76defebc0
> 
> I'm not sure what is special about `pyright`. Partially the content of the output, partially the timing I guess. But as long as it finds enough errors, this happens when I pipe the output to `head`.
> 
> (both of these machines are configured via some common code so if others can't replicate it this way then perhaps there's something special about [my config](https://github.com/MatrixManAtYrService/configs/blob/2d820cf2b752117d8a0ba67acf351cb2a4246b76/home-manager/wezterm-config.nix#L6))

I can reliably reproduce the issue using your `pyright | head` test. However, it is only highly reproducible when I add more lines. As you mentioned, this is likely a timing issue. So, the speed of ones computer may affect the reproducibility of this test. Four lines is not enough on my machine.

Also, thanks to this test, I can confirm the issue occurs on Foot _without_ Zellij. I can also confirm the issue occurs _without_ resizing.

Wrapping may be a factor. Or perhaps, wrapping just affects timing, and timing is the actual factor. I was occasionally able to reproduce the error with four bad lines when the terminal was small enough to wrap the four errors. When each error wrapped twice, I was able to reproduce the issue even _more_ often. However, this could simply be a matter of timing. I assume more processing is done the more wrapping needs to be done. And I _was_ able to reproduce the issue _without_ wrapping when adding more bad lines.

P.S. To make this test more reliable, you should include the `flake.lock` file.

**Comment by justinlovinger** [comment]
Using the `pyright` test, I was able to reproduce the issue on Alacritty and xterm, in addition to Foot. So, the terminal is likely not a factor.

**Comment by MatrixManAtYrService** [comment]
Thanks @justinlovinger for confirming, that's far fewer places for this bug to hide 😁 

Pyright is probably only special because it delays for a pathological amount of time before/between writes or before terminating.  Presumably we could write a nushell script which would replicate this just as well.  I imagine just operating over many parameters, each defining an amount of time to wait or an amount of output to write.  It should be easy to programmatically generate thousands of such permutations. Some will have the problem, hopefully 😅

I'll take a crack at it, but the challenge is that the code with the sleeps and writes won't know whether the problem has occurred, so I guess that means we'd need a small pile of subprocesses.  Like a sandwich where test code is the bread and a headless terminal and nushell are the contents:

1. test code generating various parameters and passing them to a subprocess (2) and checking for cursor coordinates in the output
2. something like https://github.com/andyk/ht which would pass the command through to its subprocess (3) and collect nushell's output to the terminal as a string which we can consume programmatically in the calling test (1)
3. nushell itself, under test, running a test command (4)
4. test code which writes, sleeps, writes, sleeps, exits, the length and duration of the writes and sleeps controlled by the parameters

One day I'll get tired of relaunching nushell to clear this state and I'll write this test.  But it feels a bit like a wheel reinvention to me, so I wanted to check first: Does anyone know a smarter way to do this kind of testing?

**Comment by MatrixManAtYrService** [comment]
Thanks @justinlovinger for confirming, that's far fewer places for this bug to hide 😁

I don't know why I didn't think of this earlier, but I think I have an easier way to replicate this.  I've recorded a session like this:

$ script typescript.log
$ pyright | head

This created [typescript.log](https://github.com/user-attachments/files/18392269/typescript.log).  On my machine, attempting to `cat typescript.log` in nushell leaves nushell in [a bad state](https://asciinema.org/a/OCotZqSpk5VMKafz1EJ672529).

**Comment by fdncred** [comment]
one approach is to delete one line at a time and cat the results until you see what line causes it to puke. Then restore the line and start deleting sequences one by one until you find the sequence that makes it puke.

I can see the prompt being drawn 5 times which points me a comment I've made before which is a reedline issue with the way it draws/redraws things.

**Comment by MatrixManAtYrService** [comment]
That's a good idea, thanks. I was able to get it down to just a few characters:

```
❯  hexdump bad.log
0000000 5b1b 6e36 0a0d
0000006

❯  echo 'hi
∙ there'
hi
there

❯  cat bad.log

❯  echo 'hi
∙ there'
^[[35;9Rhi
there
```

This also does the trick:
```
❯  python -c "import sys ; sys.stdout.buffer.write(b'\\x1b[6n\\r\\n')"
```

**Comment by fdncred** [comment]
Seems like if a tool is going to ask the terminal for the cursor position it should have something listening to receive the terminals response.

We have a `term query` command that can send and receive data from the terminal, although the bytes are in a different order than what hexdump prints.
```nushell
❯ "\e[6n\r\n" | into binary
Length: 6 (0x6) bytes | printable whitespace ascii_other non_ascii
00000000:   1b 5b 36 6e  0d 0a                                   •[6n__

❯ term query "\e[6n\r\n"
Length: 7 (0x7) bytes | printable whitespace ascii_other non_ascii
00000000:   1b 5b 34 39  3b 31 52                                •[49;1R
```

**Comment by MatrixManAtYrService** [comment]
I agree with your assessment.  And I didn't know about `term query` so that's handy.  But I think the core of this issue is that when a tool misbehaves in this way, nushell (or reedline I suppose) doesn't recover in preparation for the next command.  Cursor positioning stays messed up until you exit the shell.

If you do this in bash, by contrast, you'll see the cursor position response once, but subsequent commands not involving raw mode will work as expected (i.e. the prompt will be drawn in the correct place, neither creating large open gaps, nor clobbering the output of the previous command).

**Comment by fdncred** [comment]
> But I think the core of this issue is that when a tool misbehaves in this way, nushell (or reedline I suppose) doesn't recover in preparation for the next command. Cursor positioning stays messed up until you exit the shell.

Yup, I agree. It would be great to make this type of thing more recoverable in reedline or nushell. I'm not quite sure how to do that though. How do you detect when and external command is misbehaving? Do we need to monitor stdin or something?

**Comment by MatrixManAtYrService** [comment]
I imagine there's a loop somewhere:

1. save state (both internal nushell, and terminal flags)
2. run command
3. restore state

Could be that 3 is incomplete and we don't need to detect anything, just recover better whether or not it happened 🤷‍♂️ .

**Comment by metaspace** [comment]
I have this issue in NixOS as well:

![Image](https://github.com/user-attachments/assets/964ce3c6-c683-44ce-b798-552661919809)

Nushell 0.100.0
Starship 1.21.1
Zellij 0.41.1
Alacritty 0.14.0

It does not appear to be a problem outside Zellij, but the problem does not manifest with other shells inside Zellij with the same Starship configuration.

**Comment by justinlovinger** [comment]
It seems the issue can be temporarily fixed by running a command that sets cursor-position. `ncpamixer` works for this.

**Comment by justinlovinger** [comment]
Continuing from the above, I noticed an oddity with using `ncpamixer` to temporarily fix this issue. The volume of the first program will change as soon as `ncpamixer` is run. I think Nushell is taking the incorrect cursor-position commands and propagating them to `ncpamixer`. Presumably, other terminal-based programs will have the same issue with Nushell.

**Comment by fdncred** [comment]
If someone wants to hack at fixing this, one way may be to inspect what happens after run-external completes. If we could do some type of reset at those points, maybe it would be helpful. I did this type of thing for Windows long ago with `enable_vt_processing()`. Maybe something like that needs to happen for all operating systems and not just windows?

**Comment by Arthapz** [comment]
same problem here (nushell + zellij + starship with rio term on linux)

![Image](https://github.com/user-attachments/assets/906d50c0-0562-44e9-a189-76f4a114771b)

**Comment by Arthapz** [comment]
> It seems the issue can be temporarily fixed by running a command that sets cursor-position. `ncpamixer` works for this.

and running ncpamixer doesn't fix it on my machine :/

**Comment by Bahex** [comment]
As a temporary solution, running this should consume the extra cursor position response and fix the issue.
```nushell
term query "" -t "R"
```

**Comment by overby-me** [comment]
@Bahex it doesn't seem to work for me: (It is just stuck)

```bash
 ❯ : term query "" -t "R"
^[[13;35R
```

**Comment by Arthapz** [comment]
> [@Bahex](https://github.com/Bahex) it doesn't seem to work for me: (It is just stuck)
> 
>  ❯ : term query "" -t "R"
> ^[[13;35R

same

**Comment by KaiSforza** [comment]
I have found that using certain curses applications seems to fix this. They seem to swallow any extra characters and give me back a normal terminal. Now that the bug is reproducible, I've tested a few applications that clean up the extra characters. One that seems to be very nice is just `vim -c ':q!'`, which should grab that extra `\x1b[<x>;<y>R` output without stealing the screen or anything. It won't give back the scrollback from the previous commands that were truncted/cut off, but it should get rid of the issue for now.

Also, here's the issue for reedline, [reedline#860](https://github.com/nushell/reedline/issues/860). The `basic` terminal in there is also affected by it, so it's either reedline or some library below it that's causing this issue.

**Comment by KaiSforza** [comment]
I posted this in the discord, but for nushell this might be a temporary solution:

> Setting `$env.config.hooks.display_output` to this closure seems to properly remove the extra cursor position from the actual output.  I'll have to keep testing it out with stuff, but it seems to properly render the output (ie, links for files in `ls` and such), allow piping between commands to still be correct, and remove the extra `ansi size` from just the output. If anyone has ideas on how to make this faster or better or if there's something I've just missed, please let me know!
>
> ```
> $env.config.hooks.display_output = {||
>     if (term size).columns >= 100 {
>         table -e
>     } else {
>         table
>     }
>     | if (($in | describe) =~ "^string(| .*)") and ($in | str contains (ansi cursor_position)) {
>         str replace --no-expand --all (ansi cursor_position) ""
>     } else {
>         print -n --raw $in
>     }
> }
> ```

<img width="1016" height="530" alt="Image" src="https://github.com/user-attachments/assets/f652d6de-6ec5-430c-abe0-153b220194b1" />
